data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.processor.keyvalue;
+
+import com.amazon.dataprepper.metrics.PluginMetrics;
+import com.amazon.dataprepper.model.annotations.DataPrepperPlugin;
+import com.amazon.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.processor.AbstractProcessor;
+import com.amazon.dataprepper.model.processor.Processor;
+import com.amazon.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+@DataPrepperPlugin(name = "kv", pluginType = Processor.class, pluginConfigurationType = KeyValueProcessorConfig.class)
+public class KeyValueProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(KeyValueProcessor.class);
+
+    private final KeyValueProcessorConfig keyValueProcessorConfig;
+
+    @DataPrepperPluginConstructor
+    public KeyValueProcessor(final PluginMetrics pluginMetrics, final KeyValueProcessorConfig keyValueProcessorConfig) {
+        super(pluginMetrics);
+        this.keyValueProcessorConfig = keyValueProcessorConfig;
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        for(final Record<Event> record : records) {
+            final Map<String, Object> parsedMap = new TreeMap<>();
+            final Event recordEvent = record.getData();
+
+            final String groupsRaw = recordEvent.get(keyValueProcessorConfig.getSource(), String.class);
+            final String[] groups = groupsRaw.split(keyValueProcessorConfig.getFieldDelimiterRegex(), 0);
+            for(final String group : groups) {
+                final String[] terms = group.split(keyValueProcessorConfig.getKeyValueDelimiterRegex(), 2);
```

Comment:
I think we should look at alternatives to not use `String.split`. This method will be called a lot in this processor, and under the hood it will compile the regex patterns every single time it is called. Ideally, we should compile all regex patterns on construction of the processor.

Reply:
This should work: https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#split%28java.lang.CharSequence%29

Yes, we should pre-compile all the regex strings in the constructor.